### PR TITLE
Feature/jenkins -- addlinks

### PR DESCRIPTION
**Pull request by cookersjs**
Jenkins commit for addlinks. Addlinks is now a submodule in the data-release-pipeline repository after this change. Aside from the additon of the jenkinsfile, I changed some config values in create-references-context.xml and file-download-context.xml, and added in the link checking to the former context file. Hopefully there isn't a major conflict anywhere, but I will resolve it if so.

Thanks!

### CODE REVIEW

File: src/main/resources/file-download-context.xml [hunk 1]
```diff
@@ -40,30 +40,34 @@
...
+		<value>IntEnz</value>		
+		<value>UniProtToENSEMBL</value>
 		<value>EnsemblToALL</value>
-		<value>UniProtToOMIM</value>
+		<value>OpenTargets</value>
+		<value>FlyBaseToUniprotReferenceDNASequence</value>
+		<value>CTD</value>
 		<value>UniProtToEntrez_Gene</value>
-		<value>UniProtToPDB</value>
+		<value>UniProtToUCSC</value>
```
Comment (jweiser) [inline src/main/resources/file-download-context.xml]: Just curious:  Why have `UniProtToENSEmBL` and `EntrezGeneToUniprotReference` been removed and `UniProtToUCSC` added?

File: Jenkinsfile-downloads [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+			steps{
+				script{
+					def configStatusUrl = httpRequest authentication: 'jenkinsKey', url: "${env.JENKINS_JOB_URL}/job/${env.RELEASE_NUMBER}/job/ConfirmReleaseConfigs/lastBuild/api/json"
+					def configStatusJson = new JsonSlurper().parseText(configStatusUrl.getContent())
+					if(configStatusJson['result'] != "SUCCESS"){
+						error("Most recent ConfirmReleaseConfigs build status: " + configStatusJson['result'] + ". Please complete a successful build.")
+					}
+			    	}	
+		    	}
+	    	}
```
Comment (jweiser) [inline Jenkinsfile-downloads]: Is it possible to add some inline documentation on what this step is checking overall?

File: Jenkinsfile-downloads [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import groovy.json.JsonSlurper
+
+pipeline{
+	agent any
+
+	stages{
+		stage('Check upstream builds succeeded'){
+			steps{
+				script{
+					def configStatusUrl = httpRequest authentication: 'jenkinsKey', url: "${env.JENKINS_JOB_URL}/job/${env.RELEASE_NUMBER}/job/ConfirmReleaseConfigs/lastBuild/api/json"
```
Comment (jweiser) [inline Jenkinsfile-downloads]: A comment for this line would be great.  It is a bit cryptic to me.

File: Jenkinsfile-downloads [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+						sh "mvn clean package -DskipTests"
+                    			}
+          	    		}
+            		}		
+        	}
+		stage('Main: Download AddLinks files'){
+			steps{
+				script{
+					dir('add-links'){
+						withCredentials([file(credentialsId: 'Config', variable: 'FILE')]){
```
Comment (jweiser) [inline Jenkinsfile-downloads]: Could the `FILE` variable be more descriptive?

File: Jenkinsfile-downloads [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+							sh "cp $FILE src/main/resources/auth.properties"
+							sh "ln -sf src/main/resources resources"
+							sh "rm -f src/main/resources/db.properties"
```
Comment (jweiser) [inline Jenkinsfile-downloads]: Why is `db.properties` removed here?  Is it handled by Jenkins master config file?

File: Jenkinsfile-insertions [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+					}
+				}
+			}
+		}
+		stage('Post: Backup DB'){
+			steps{
+				script{
+					dir('add-links'){
+						withCredentials([usernamePassword(credentialsId: 'mySQLUsernamePassword', passwordVariable: 'pass', usernameVariable: 'user')]){
+							sh "mysqldump -u$user -p$pass ${env.RELEASE_CURRENT} > ${env.RELEASE_CURRENT}_${env.RELEASE_NUMBER}_after_addlinks.dump"
```
Comment (jweiser) [inline Jenkinsfile-insertions]: I remember we talked about it, but I forget what our conclusion was on giving `mysql` and `mysqldump` a password outside of the command-line?  Was it necessary/doable?

File: Jenkinsfile-insertions [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		}
+		stage('Post:Archive logs and backups'){
+			steps{
+				script{
+					dir('add-links'){
+						sh "mkdir -p archive/${env.RELEASE_NUMBER}/logs"
+						sh "mv --backup=numbered *_${env.RELEASE_NUMBER}_*.dump.gz archive/${env.RELEASE_NUMBER}/"
+						sh "gzip logs/*.log"
+						sh "gzip logs/retrievers/*"
+						sh "gzip logs/*.log logs/retrievers/* logs/file-processors/* logs/refCreators/*"
```
Comment (jweiser) [inline Jenkinsfile-insertions]: In the old pipeline, `gzip` uses the -q option to suppress errors if you think it's useful but with Jenkins logging everything it might be good to see a failure?

File: Jenkinsfile-insertions [hunk 8]
```diff
@@ -0,0 +1,88 @@
+import groovy.json.JsonSlurper
+
+pipeline{
+	agent any
+
+	stages{
+		stage('Check upstream builds succeeded'){
```
Comment (jweiser) [inline Jenkinsfile-insertions]: Maybe this could read "Check AddLinks Download and Orthoinference upstream builds succeeded"?

File: Jenkinsfile-insertions [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+			steps{
+				script{
+					def addLinksDownloadStatusUrl = httpRequest authentication: 'jenkinsKey', url: "${env.JENKINS_JOB_URL}/job/${env.RELEASE_NUMBER}/job/AddLinks-Download/lastBuild/api/json"
+					def addLinksDownloadJson = new JsonSlurper().parseText(addLinksDownloadUrl.getContent())
+					if(addLinksDownloadJson['result'] != "SUCCESS"){
+						error("Most recent AddLinks-Download build status: " + addLinksDownloadJson['result'] + ". Please complete a successful build.")
+					}
+			    	}
+				script{
+					def orthoinferenceDownloadStatusUrl = httpRequest authentication: 'jenkinsKey', url: "${env.JENKINS_JOB_URL}/job/${env.RELEASE_NUMBER}/job/Orthoinference/lastBuild/api/json"
```
Comment (jweiser) [inline Jenkinsfile-insertions]: Should this be `def orthoinferenceStatusUrl`?

File: Jenkinsfile-insertions [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+						sh "mvn clean package -DskipTests"
+                    			}
+          	    		}
+            		}
+        	}
+		stage('Main: AddLinks Insertions'){
+			steps{
+				script{
+					dir('add-links'){
+						withCredentials([file(credentialsId: 'Config', variable: 'FILE')]){
```
Comment (jweiser) [inline Jenkinsfile-insertions]: Similar to above - a change to the `FILE` variable name to be more specific?

File: src/main/resources/file-download-context.xml (hunk 1 above)
Comment (cookersjs) [reply src/main/resources/file-download-context.xml]: I wasn't the one who actually made that change. Basically, 'file-download-context.xml' and 'create-references-context.xml' were older config files that were originally used to run either the download or insertion steps in add-links. They're basically the download/insertion parts of 'application-context.xml', split up. Since we've used the Perl wrapper to do this over the last year to comment out application-context, these two files became out of sync due to disuse. For Jenkins, I am reviving their use and I just copied the pertinent parts from application-context to update them.

In short, this is actually a change that probably happened to add-links in the last year or so, but I don't know for sure

File: Jenkinsfile-downloads (hunk 5 above)
Comment (cookersjs) [reply Jenkinsfile-downloads]: The addlinks process basically searches in the resources folder for *.properties, and then creates a kind of mega properties file to be used by the program. `db.properties` is a default file that has most of its attributes blank. With Jenkins now supplying the config file, the path of least resistance was to have Jenkins replace one of the two .properties files (auth.properties or db.properties), and then remove the other file so that duplicate key-value pairs don't pollute the process

File: Jenkinsfile-insertions (hunk 6 above)
Comment (cookersjs) [reply Jenkinsfile-insertions]: I am sure its doable, but this is probably more easily tackled when we actually deploy this on the release server, in case there are any differences in how the Jenkins user is utilized there than on my computer. For now, Jenkins is very good for making sure credentialed values show up as asterisks. Definitely can consider the use of the `my.cnf` file once its in place, though I don't know if having to switch between the local release server and curator servers are going to make things difficult, a la the `pass2production.sh` issue we had for release 70

File: Jenkinsfile-insertions (hunk 7 above)
Comment (cookersjs) [reply Jenkinsfile-insertions]: I'm a little confused at this comment. Is this more of a thought to be aware of or are you advocating for the use of this flag?

File: Jenkinsfile-insertions (hunk 9 above)
Comment (cookersjs) [reply Jenkinsfile-insertions]: Yes it should be, you caught my copy-pasting act

File: Jenkinsfile-insertions (hunk 10 above)
Comment (cookersjs) [reply Jenkinsfile-insertions]: Yea, I thing in some files I have the variable name changed to ConfigFile. I'll make that uniform across the Jenkinsfiles

**Comment by SolomonShorser-OICR** [comment]
I notice that many of the steps have `script` blocks that just call shell commands, one at a time. Are there no Jenkins plugins/extensions to gzip, ln, etc...?

File: src/main/resources/create-references-context.xml [hunk 11]
```diff
@@ -25,11 +25,11 @@
 	public MySQLAdaptor(String host,String database,String username,String password,int port)
 	-->
 	<bean id="dbAdapter" name="dbAdapter" class="org.gk.persistence.MySQLAdaptor" scope="singleton">
-		<constructor-arg name="host" index="0" value="${database.host}"/>
-		<constructor-arg name="database" index="1" value="${database.name}"/>
-		<constructor-arg name="username" index="2" value="${database.user}"/>
-		<constructor-arg name="password" index="3" value="${database.password}"/>
-		<constructor-arg name="port" index="4" value="${database.port}"/>
+		<constructor-arg name="host" index="0" value="${release.database.host}"/>
```
Comment (SolomonShorser-OICR) [inline src/main/resources/create-references-context.xml]: A note: The other properties files and Spring context files in this project that _currently_ do not have a `release.` prefix may cause confusion to future developers. I recommend changing other properties/context files to match this new naming convention.

File: src/main/resources/file-download-context.xml (hunk 1 above)
Comment (SolomonShorser-OICR) [reply src/main/resources/file-download-context.xml]: UniProtToENSEMBL was moved from line 56 to 44, not removed. From what I can tell, `EntrezGeneToUniprotReference` was never used, however file-download-context.xml is probably old enough that the reference to that bean still existed. It has not been in the main application-context.xml for a while.

File: Jenkinsfile-insertions (hunk 6 above)
Comment (jweiser) [reply Jenkinsfile-insertions]: Sounds good - we can work on it in a new PR, if necessary, when deploying to the release server.

File: Jenkinsfile-insertions (hunk 7 above)
Comment (jweiser) [reply Jenkinsfile-insertions]: Just mentioning in case you thought it was useful - not advocating for it.

File: src/main/resources/create-references-context.xml (hunk 11 above)
Comment (cookersjs) [reply src/main/resources/create-references-context.xml]: Yes, as discussed I will be making sure that the changed configs are seen in the other config file(s). Thanks!

Comment (cookersjs) [reply src/main/resources/create-references-context.xml]: This has now been completed. Now all 3 context files (application, create-references, and file-download) have matching configurations

**Comment by cookersjs** [comment]
@SolomonShorser-OICR @jweiser Requested changes have been done for the AddLinks Jenkins branch. Please re-review and let me know if other changes are needed.

File: Jenkinsfile-insertions [hunk 12]
```diff
@@ -84,5 +84,14 @@ pipeline{
...
+}
+
+// Utility function that checks upstream builds of this project were successfully built.
+def checkUpstreamBuildsSucceeded(String stepName) {
+	def statusUrl = httpRequest authentication: 'jenkinsKey', url: "${env.JENKINS_JOB_URL}/job/${env.RELEASE_NUMBER}/job/$stepName/lastBuild/api/json"
+	def statusJson = new JsonSlurper().parseText(statusUrl.getContent())
+	if(statusJson['result'] != "SUCCESS"){
+		error("Most recent $stepName build status: " + statusJson['result'])
+	}
 }
```
Comment (jweiser) [inline Jenkinsfile-insertions]: This is really good :).  I was wondering since this check is done in many release steps for the Jenkins pipeline, is there a way to make it a utility function at higher level so that it can be imported and reused across steps?  

If so, it'd be good to make a GitHub issue to do that for a new PR.

File: Jenkinsfile-downloads [hunk 13]
```diff
@@ -1,43 +1,49 @@
...
 				script{
 					dir('add-links'){
-						withCredentials([file(credentialsId: 'Config', variable: 'FILE')]){
-							sh "cp $FILE src/main/resources/auth.properties"
+						withCredentials([file(credentialsId: 'Config', variable: 'ConfigFile')]){
+							sh "cp $ConfigFile src/main/resources/auth.properties"
 							sh "ln -sf src/main/resources resources"
 							sh "rm -f src/main/resources/db.properties"
-					    		sh """java -cp "\$(pwd)/resources" -Dconfig.location=\$(pwd)/resources/addlinks.properties -Dlog4j.configurationFile=\$(pwd)/resources/log4j2.xml -Xmx${env.JAVA_MEM_MAX}m -jar target/AddLinks-${env.ADDLINKS_VERSION}.jar file://\$(pwd)/resources/create-references-context.xml"""							
+							sh """java -cp "\$(pwd)/resources" -Dconfig.location=\$(pwd)/resources/addlinks.properties -Dlog4j.configurationFile=\$(pwd)/resources/log4j2.xml -Xmx${env.JAVA_MEM_MAX}m -jar target/AddLinks-${env.ADDLINKS_VERSION}.jar file://\$(pwd)/resources/create-references-context.xml"""
```
Comment (jweiser) [inline Jenkinsfile-downloads]: Just curious - why the triple quotes around the java command?

File: Jenkinsfile-insertions [hunk 14]
```diff
@@ -1,75 +1,75 @@
...
+		stage('Post: Backup release_current'){
 			steps{
 				script{
 					dir('add-links'){
 						withCredentials([usernamePassword(credentialsId: 'mySQLUsernamePassword', passwordVariable: 'pass', usernameVariable: 'user')]){
-							sh "mysqldump -u$user -p$pass ${env.RELEASE_CURRENT} > ${env.RELEASE_CURRENT}_${env.RELEASE_NUMBER}_after_addlinks.dump"
-							sh "gzip ${env.RELEASE_CURRENT}_${env.RELEASE_NUMBER}_after_addlinks.dump"
+							def release_current_after_addlinks_dump = "${env.RELEASE_CURRENT}_${env.RELEASE_NUMBER}_after_addlinks.dump"
+							sh "mysqldump -u$user -p$pass ${env.RELEASE_CURRENT} > $release_current_after_addlinks_dump"
+							sh "gzip $release_current_after_addlinks_dump"
```
Comment (jweiser) [inline Jenkinsfile-insertions]: A project level utility function to backup and gzip databases would be good similar to https://github.com/reactome/add-links/pull/127/files/a6d3d0c59207884075bf057f022b07af5b784552..3809d4507f00851f8b0ec095e39200fccd4c4d54#r333680454

File: Jenkinsfile-insertions (hunk 12 above)
Comment (cookersjs) [reply Jenkinsfile-insertions]: To the best of my knowledge, we'd have to do it through a bash script, similar to the 'confirmJenkinsConfigs' in the scripts folder. We'd have to supply it the URL and authentication information for accessing the Jenkins API. I don't see why that wouldn't be possible so I'll create an issue

Comment (cookersjs) [reply Jenkinsfile-insertions]: Created one in data-release-pipeline repo: https://github.com/reactome/data-release-pipeline/issues/100

File: Jenkinsfile-downloads (hunk 13 above)
Comment (cookersjs) [reply Jenkinsfile-downloads]: After 'java -cp' the following argument ("$(pwd)/resources") has always had double quotes. In order to keep that, and allow the introduction of jenkins variables I couldn't use single quotes since they force the literal (and they also couldn't be used for the $pwd/resources argument, I believe)
